File: core/src/test/java/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraintsTest.java
package fr.sncf.osrd.api.pathfinding.constraints;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.common.collect.Range;
import fr.sncf.osrd.infra.api.Direction;
import fr.sncf.osrd.infra.api.tracks.undirected.NeutralSection;
import fr.sncf.osrd.stdcm.DummyRouteGraphBuilder;
import fr.sncf.osrd.train.TestTrains;
import fr.sncf.osrd.utils.graph.Pathfinding;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import java.util.List;
import java.util.stream.Stream;

public class ElectrificationConstraintsTest {
    static Stream<Arguments> testNeutralSectionArgs() {
        var fullCatenaryRanges = List.of(Range.closed(0., 1000.));
        var partialCatenaryRanges = List.of(Range.closed(0., 100.), Range.closed(900., 1000.));
        var noCatenaryRanges = List.of();
        return Stream.of(
                // Catenary ranges, with neutral section, neutral section direction
                Arguments.of(fullCatenaryRanges, true, Direction.FORWARD),
                Arguments.of(fullCatenaryRanges, true, Direction.BACKWARD),
                Arguments.of(fullCatenaryRanges, false, null),
                Arguments.of(partialCatenaryRanges, true, Direction.FORWARD),
                Arguments.of(partialCatenaryRanges, true, Direction.BACKWARD),
                Arguments.of(partialCatenaryRanges, false, null),
                Arguments.of(noCatenaryRanges, true, Direction.FORWARD),
                Arguments.of(noCatenaryRanges, false, null)
        );
    }

    @ParameterizedTest
    @MethodSource("testNeutralSectionArgs")
    public void testNeutralSectionAndCatenaryBlockingRanges(List<Range<Double>> catenaryRanges,
                                                            boolean withNeutralSection,
                                                            Direction neutralSectionDirection) {

        var route = new DummyRouteGraphBuilder.DummyRoute("route", 1000, null, null, null);

        assert TestTrains.FAST_ELECTRIC_TRAIN.getModeNames().contains("25000");
        var catenaryLength = 0.;
        for (var catenaryRange : catenaryRanges) {
            route.getTrackRanges().get(0).track.getEdge().getVoltages().put(catenaryRange, "25000");
            catenaryLength += catenaryRange.upperEndpoint() - catenaryRange.lowerEndpoint();
        }

        if (withNeutralSection) {
            var neutralSections = route.getTrackRanges().get(0).track.getEdge().getNeutralSections(
                    neutralSectionDirection);
            neutralSections.put(Range.closedOpen(100., 950.), new NeutralSection(false));
        }

        var blockedRanges = new ElectrificationConstraints(List.of(TestTrains.FAST_ELECTRIC_TRAIN)).apply(route);

        if (catenaryLength > 900) {
            assertEquals(0, blockedRanges.size());
        } else if (catenaryLength > 100) {
            if (withNeutralSection && neutralSectionDirection == Direction.FORWARD) {
                assertEquals(0, blockedRanges.size());
            } else {
                assertEquals(1, blockedRanges.size());
                assertTrue(blockedRanges.contains(new Pathfinding.Range(100., 900.)));
            }
        } else {
            if (withNeutralSection && neutralSectionDirection == Direction.FORWARD) {
                assertEquals(2, blockedRanges.size());
                assertTrue(blockedRanges.contains(new Pathfinding.Range(0., 100.)));
                assertTrue(blockedRanges.contains(new Pathfinding.Range(950., 1000.)));
            } else {
                assertEquals(1, blockedRanges.size());
                assertTrue(blockedRanges.contains(new Pathfinding.Range(0., 1000.)));
            }
        }
    }
}

File: core/src/test/kotlin/fr/sncf/osrd/pathfinding/constraints/ElectrificationConstraintsTest.kt
package fr.sncf.osrd.pathfinding.constraints

import fr.sncf.osrd.pathfinding.Pathfinding
import fr.sncf.osrd.sim_infra.api.Block
import fr.sncf.osrd.sim_infra.api.BlockId
import fr.sncf.osrd.sim_infra.api.TrackChunk
import fr.sncf.osrd.sim_infra.api.TrackChunkId
import fr.sncf.osrd.train.TestTrains
import fr.sncf.osrd.utils.Helpers
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import java.util.stream.Stream
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ElectrificationConstraintsTest {
    private var electrificationConstraints: ElectrificationConstraints? = null
    private var chunk0Length: Offset<TrackChunk> = Offset(0.meters)

    @BeforeAll
    fun setUp() {
        val infra = Helpers.smallInfra
        electrificationConstraints =
            ElectrificationConstraints(
                infra.blockInfra,
                infra.rawInfra,
                TestTrains.FAST_ELECTRIC_TRAIN.modeNames,
            )
        chunk0Length = infra.rawInfra.getTrackChunkLength(TrackChunkId(0U))
    }

    @ParameterizedTest
    @MethodSource("testDeadSectionArgs")
    fun testDeadSectionAndElectrificationBlockedRanges(
        blockId: BlockId,
        expectedBlockedRanges: Collection<Pathfinding.Range<Block>>,
    ) {
        val blockedRanges = electrificationConstraints!!.apply(blockId)
        AssertionsForClassTypes.assertThat(blockedRanges).isEqualTo(expectedBlockedRanges)
    }

    fun testDeadSectionArgs(): Stream<Arguments> {
        return Stream.of( // No corresponding electrification ranges without dead sections
            Arguments.of(
                0,
                mutableSetOf(Pathfinding.Range(Offset(0.meters), chunk0Length)),
            ), // Partially corresponding electrification ranges with dead
            // section
            Arguments.of(
                1,
                mutableSetOf(Pathfinding.Range(Offset<Block>(0.meters), Offset(140.meters))),
            ), // Fully corresponding electrification ranges without dead
            // sections
            Arguments.of(2, HashSet<Any>()),
        )
    }
}
